src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeHelpers.cs

```diff
@@ -109,6 +110,35 @@ internal static bool IsPrimitiveType(this CorElementType et)
             // COR_ELEMENT_TYPE_I1,I2,I4,I8,U1,U2,U4,U8,R4,R8,I,U,CHAR,BOOLEAN
             => ((1 << (int)et) & 0b_0011_0000_0000_0011_1111_1111_1100) != 0;
 
+        private static ReadOnlySpan<short> PrimitiveWidenTable =>
+            [
+                0x00,      // ELEMENT_TYPE_END
+                0x00,      // ELEMENT_TYPE_VOID
+                0x0004,    // ELEMENT_TYPE_BOOLEAN
+                0x3F88,    // ELEMENT_TYPE_CHAR (W = U2, CHAR, I4, U4, I8, U8, R4, R8) (U2 == Char)
+                0x3550,    // ELEMENT_TYPE_I1   (W = I1, I2, I4, I8, R4, R8)
+                0x3FE8,    // ELEMENT_TYPE_U1   (W = CHAR, U1, I2, U2, I4, U4, I8, U8, R4, R8)
+                0x3540,    // ELEMENT_TYPE_I2   (W = I2, I4, I8, R4, R8)
+                0x3F88,    // ELEMENT_TYPE_U2   (W = U2, CHAR, I4, U4, I8, U8, R4, R8)
+                0x3500,    // ELEMENT_TYPE_I4   (W = I4, I8, R4, R8)
+                0x3E00,    // ELEMENT_TYPE_U4   (W = U4, I8, R4, R8)
+                0x3400,    // ELEMENT_TYPE_I8   (W = I8, R4, R8)
+                0x3800,    // ELEMENT_TYPE_U8   (W = U8, R4, R8)
+                0x3000,    // ELEMENT_TYPE_R4   (W = R4, R8)
+                0x2000,    // ELEMENT_TYPE_R8   (W = R8)
+            ];
+
+        internal static bool CanPrimitiveWiden(CorElementType srcET, CorElementType dstET)
+        {
+            Debug.Assert(srcET.IsPrimitiveType() && dstET.IsPrimitiveType());
+            if ((int)srcET >= PrimitiveWidenTable.Length)
+            {
+                // I or U
+                return srcET == dstET;
+            }
+            return (PrimitiveWidenTable[(int)srcET] & (1 << (int)dstET)) != 0;
+        }
```

Comment:
These are really sharable with NativeAOT, but `EETypeElementType` uses different names and values.
Can we reduce the gaps here? Are these values already hardcoded by diagnostic tools?

Reply:
Yes, it would be nice to figure a way to dedup these. It should be a separate change.

Reply:
(The native AOT encoding of the type kinds is more compact. It saves some bits in MethodTable.)

Reply:
> Are these values already hardcoded by diagnostic tools?

We can fix diagnostic tools as needed.

Reply:
`CorElementType` was shipped in `corhdr.h` and included in Windows SDK and mingw SDK (I found it in Debian source tree). It can be used by many external tools. We really have no chance to make breaking change to it.

Reply:
Right, `CorElementType` is used by ECMA-335 metadata encoding. 

Nothing says that the runtime has to use the exact same enum internally everywhere to represent type kinds.

For example, `CorInfoType` that is used by the JIT and JIT/EE interface is enum like that. It is different from `CorElementType`. Managed `TypeCode` is another example of a type kind enum that is different from `CorElementType`